File: leetcode/medium/Q96_UniqueBinarySearchTrees.py
class Solution:
    def numTrees(self, n: int) -> int:
        answer = [0] * (n+2)
        answer[0] = 1
        answer[1] = 1

        for i in range(2, n+1):
            answer[i] = 0
            for j in range(i, 0, -1):
                answer[i] += (answer[j-1] * answer[i - j])

        return answer[n]

s = Solution()
a = s.numTrees(3)
print(a)

File: src/medium/Q96_UniqueBinarySearchTrees.java
package medium;

public class Q96_UniqueBinarySearchTrees {
    class Solution {
        public int numTrees(int n) {
            int[] cat = new int[n+1];
            cat[0] = 1;
            cat[1] = 1;
            for (int i = 2; i <= n; i++) {
                for (int j = 1; j <= i; j++) {
                    cat[i] += cat[j-1] * cat[i-j];
                }
            }
            return cat[n];
        }
    }
}
